### PR TITLE
Add ldftn and DelegateShim

### PR DESCRIPTION
**Pull request by BadRyuner**
- Adds support for Ldftn OpCode
- Adds a minimal wrapper for Delegates (.ctor & invoke) + mini test for this thing

### CODE REVIEW

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        // skip 1 for delegate "this"
+        for (var i = 1; i < arguments.Count; i++)
+            stack.Push(arguments[i], method.Signature!.ParameterTypes[i]);
+
+        var result = vm.Dispatcher.Dispatch(context, new(CilOpCodes.Call, method));
+
+        if (!result.IsSuccess)
+            return InvocationResult.Exception(result.ExceptionObject);
+
+        return InvocationResult.StepOver(null);
```
Comment (Washi1337) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs:98]: I feel this is problematic or at the very least inaccurate:

Firstly, this goes through CIL instruction dispatching when in real processes this does not happen. I feel we should feed the invocation back to the VM's method invoker and return its invocation result. This also avoids a lot of unnecessary marshalling to stack slots and back.

Additionally, skipping the call with `StepOver(null)` results in the `Invoke` method call actually not being registered as a call in the callstack of the thread. This isn't usually a problem for other shims (such as methods in `Unsafe`) as we intend to immediately return from these methods anyway. However, depending on the VM's method invoker, we may actually need to step into the method referenced by the delegate. Not having `Invoke` in the callstack then is not just inaccurate emulation, but can also be detected by the callee when we get to actually emulating call stacks in `StackTrace` and `Exception` objects.

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/IFunctionMarshaller.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    /// </summary>
+    CilVirtualMachine Machine
+    {
+        get;
+    }
+
+    /// <summary>
+    /// Returns pointer for method descriptor
+    /// </summary>
+    nuint GetFunctionPointer(IMethodDescriptor method);
```
Comment (Washi1337) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/IFunctionMarshaller.cs:20]: We cannot use `nint` or `nuint` for pointer-typed values, as the host process may have a different bitness than the emulated process. This is why `ulong` is used for addresses throughout the library instead.

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/IFunctionMarshaller.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+using AsmResolver.DotNet;
+
+namespace Echo.Platforms.AsmResolver.Emulation;
+/// <summary>
+/// Provides methods for marshalling functions into pointers
+/// </summary>
+public interface IFunctionMarshaller
```
Comment (Washi1337) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/IFunctionMarshaller.cs:7]: Do we actually need a separate `IFunctionMarshaller` interface and `FunctionMarshaller` implementation class?

[ClrMockMemory](https://github.com/Washi1337/Echo/blob/master/src/Platforms/Echo.Platforms.AsmResolver/Emulation/Runtime/ClrMockMemory.cs) is already HLE'ing various internal structures of the CLR and mapping them to virtual addresses using `GenericMockMemory<T>`. I feel the proposed implementation is kind of reinventing the wheel for something that could simply be a `GenericMockMemory<IMethodDescriptor> MethodEntryPoints { get; }` in `ClrMockMemory`.

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/FunctionMarshaller.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+using System.Linq;
+using AsmResolver.DotNet;
+
+namespace Echo.Platforms.AsmResolver.Emulation;
+
+/// <inheritdoc/>
+public class FunctionMarshaller : IFunctionMarshaller
+{
+    private readonly Dictionary<nuint, IMethodDescriptor> mappedPointers = new();
+    private nuint pointer = 13;
```
Comment (Washi1337) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/FunctionMarshaller.cs:12]: This is not a valid pointer in the VM's virtual memory.

**Review by Washi1337 — CHANGES_REQUESTED** [review]
This is a good idea and a nice start, however, we need to be super careful here with the implementation details to avoid any inaccuracies that can potentially lead to problems later, or be detected by the emulated process.

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs (hunk 1 above)
Comment (BadRyuner) [reply src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs:89]: I have two ideas:
1) A FakeCallFrame that will only show that this method was called. (lots of virtual and overload ctrl+c ctrl+v)
2) Create (if MethodDefinition.CilMethodBody == null) a CilMethodBody for Invoke method, which will be like: optional(ldarg_0 -> ldfld _target) -> ldarg [0..N] -> ldarg_0 -> ldfld _methodptr -> calli -> ret. And return StepIn or how it works.

Comment (Washi1337) [reply src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs:89]: Hmm this is a tough choice. Neither option is great, and seems to highlight a fundamental limitation in the current invoker API. It makes me wonder if `InvocationResult` should get another result type, something like `InvocationResultType::FullyHandled` to allow method invokers to fully override the call behavior for extreme cases like these.

With such a result type, it then may make the most sense to let the delegate invoker push a normal `CallFrame` for the `Invoke` method, and then off-load to the VM's method invoker.
- If that results in a step-in, we keep the `Invoke` call frame on the stack before returning `FullyHandled`.
- For any other result, we can pop the `Invoke` call frame again before returning said result.

The only open question then is how to handle the final return from the `Invoke` method to the caller.

Comment (BadRyuner) [reply src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs]: Something like that? [commit](https://github.com/Washi1337/Echo/pull/140/commits/b2aebc0c2c8370c5615b403398e79947ad79e9e4)

Comment (Washi1337) [reply src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs]: I like the idea of `IsTrampoline`, that works out nicely.

However like I said before, I really think `DelegateInvoker` should not be the one setting up the underlying call itself and instead use `context.Machine.Invoker` to do this. This way users can still customize (i.e. shim) calls that are behind delegates.

Comment (BadRyuner) [reply src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs]: I forgot about Invoker. I made a new commit.

File: test/Platforms/Echo.Platforms.AsmResolver.Tests/Emulation/CilVirtualMachineTest.cs [hunk 5]
```diff
@@ -662,5 +662,59 @@ public void StepPrefixedInstructionsShouldStepOverAllInstructions()
             _mainThread.Step();
             Assert.Equal(instructions[3].Offset, frame.ProgramCounter);
         }
+        
+        [Fact]
+        public void CallDelegate()
```
Comment (Washi1337) [inline test/Platforms/Echo.Platforms.AsmResolver.Tests/Emulation/CilVirtualMachineTest.cs:667]: Is there a way to clean up this test a bit, such that it is a bit more clear what exactly is tested for? I am having a bit of a hard time understanding exactly what is happening in this test. 

You can be explicit and repetitive in unit tests. The more explicit they are (especially in their asserts) the better context it gives us when something breaks in the future.

**Review by Washi1337 — COMMENTED** [review]
Apologies for the late reply, and also for not coming with the following comments earlier in the previous review, but I am noticing a few more potential edge cases that we may want to address before merging. 

Really appreciate the work you put into it btw!

File: test/Platforms/Echo.Platforms.AsmResolver.Tests/Emulation/CilVirtualMachineTest.cs (hunk 5 above)
Comment (BadRyuner) [reply test/Platforms/Echo.Platforms.AsmResolver.Tests/Emulation/CilVirtualMachineTest.cs:667]: I think it's clearer now.

**Comment by Washi1337** [comment]
Thanks a lot!
